### PR TITLE
Make arrow keys work with "Create new option" option.  

### PR DESCRIPTION
**Pull request by Barraketh**
A bit hacky, but i'm trying to minimize code changes.

### CODE REVIEW

**Comment by Barraketh** [comment]
@mikemintz @brian-element @elementsean

**Comment by mikemintz** [comment]
@Barraketh can you briefly describe the mechanism this uses to create a new option? I'm not familiar with the code base so I'm not sure what's happening.

File: src/Select.js [hunk 1]
```diff
@@ -661,11 +661,31 @@ var Select = React.createClass({
 		this.focusAdjacentOption('previous');
 	},
 
+	createNewOption() {
+		if (this.props.allowCreate && this.state.inputValue.trim()) {
+			var inputValue = this.state.inputValue;
```
Comment (brian-element) [inline src/Select.js]: I would go with `let` instead of `var` but it conflicts with the style of the rest of the code so I'm not sure.

**Comment by Barraketh** [comment]
Basically, it works like this:

The currently focused option is maintained in state.focusedOption.  There is also a state.filteredOptions variable that maintains the defined options we're working with (not entirely sure why it's called filtered)

there is a buildMenu() function which, in the process of building the open menu, used to create the "Add {}" option, create a new array which is this option + filteredOptions, and then work with that.

The key handling is done by the focusAdjacentOption function, which only works on the filteredOptions array.  I basically replicate the logic in buildMenu() in focusAdjacentOption()

File: src/Select.js [hunk 2]
```diff
@@ -661,11 +661,31 @@ var Select = React.createClass({
...
+
 	focusAdjacentOption (dir) {
 		this._focusedOptionReveal = true;
 		var ops = this.state.filteredOptions.filter(function(op) {
 			return !op.disabled;
 		});
+
+		let newOption = this.createNewOption();
+		if (newOption !== null) {
+			ops = ops.slice();
```
Comment (brian-element) [inline src/Select.js]: I feel like

```javascript
ops = [ newOption, ...ops ]
```
or

```javascript
ops = [].concat(newOption, ops)
```

works better here.

File: src/Select.js (hunk 1 above)
Comment (Barraketh) [inline src/Select.js]: yeah, I literally copied that code from buildMenu()

File: src/Select.js (hunk 2 above)
Comment (mikemintz) [inline src/Select.js]: I think the second syntax would have `ops` be doubly-nested. The first option is fine, but they may not have spread-syntax in their babel settings.

File: src/Select.js [hunk 3]
```diff
@@ -719,14 +739,9 @@ var Select = React.createClass({
 		}
 		// Add the current value to the filtered options in last resort
 		var options = this.state.filteredOptions;
-		if (this.props.allowCreate && this.state.inputValue.trim()) {
-			var inputValue = this.state.inputValue;
+		let newOption = this.createNewOption();
+		if (newOption !== null) {
 			options = options.slice();
```
Comment (brian-element) [inline src/Select.js]: Same here but now I see that increases the surface area of the change.

**Comment by brian-element** [comment]
Small comments for you to contemplate but LGTM.

**Comment by mikemintz** [comment]
It makes sense to me now I think. The key change is line 702, adding the `this.state.focusedOption.create && ops[i].create` to the condition that allows focusing an option, right?

If we're aiming to get this merged, I'd suggest changing the two `let` keywords to `var` so it's consistent with the old style. I'd also suggest splitting this into two commits: one that just refactors by adding and using the `createNewOption` method, and the other that actually changes the behavior.

Otherwise LGTM

File: src/Select.js (hunk 2 above)
Comment (brian-element) [inline src/Select.js]: Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat

> `concat` creates a new array consisting of the elements in the object on which it is called, followed in order by, for each argument, the elements of that argument (if the argument is an array) or the argument itself (if the argument is not an array).

So I think concat is meant to flatten arrays. E.g.:

```
> let a = [1, 2]
undefined
> [].concat(3, a)
[ 3, 1, 2 ]
```

Not extremely important. But it is worth knowing.

Comment (mikemintz) [inline src/Select.js]: Oh good to know! I had thought it just flattens the first argument if there's only one argument as some sort of special case.

**Pull request closed without merging**